[CI] Fixed simple typo in `doc/` to test docs-pipeline

ueqri opened the pull request:
Current docs pipeline seems to work well for push/PR which contains code changes(i.e. files excluding `doc/*` and `*.md` in root directory), according to the [Azure Pipeline Dashboard](https://dev.azure.com/PointCloudLibrary/pcl/_build?view=runs).

To test only doc changes, which will **only trigger docs-pipeline**, I fixed some simple typo in `doc/` I've found with spell check plugin open when skimming through the documentations before.

Expected behavior of the pipelines at present:
- Doc changes: format -> generate doc -> run tutorials stage
- Code changes: format -> run build -> (if build_gcc stage passed) [generate doc -> run tutorials stage]

ueqri commented:
![Screenshot_2021-04-20_21-02-04](https://user-images.githubusercontent.com/56567688/115400773-3d876700-a21c-11eb-9e8d-790fa31aa2ed.png)
![Screenshot_2021-04-20_21-11-00](https://user-images.githubusercontent.com/56567688/115401399-ee8e0180-a21c-11eb-8f0c-5acda0003951.png)

The test result looks good. 

But still a little confused about [the first](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=18755&view=results) and [the second](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=18753&view=results) pipeline in the table above. I'm not very sure why the docs pipeline was triggered twice. Maybe the reason is the commit contains both code changes and docs changes, `*.rst` changes run the docs pipeline and `*.sh` changes trigger the main build. Should I do some further job to find out the causes and fine tune the CI?

kunaltyagi commented:
Yup, that's a flaw. The CI seems to be running it on both the merge to master trigger, and the build-gcc trigger

ueqri commented:
> Yup, that's a flaw. The CI seems to be running it on both the merge to master trigger, and the build-gcc trigger

To remedy the unexpected behavior, I've introduced `git diff` in [my test branch](https://github.com/ueqri/pcl/tree/test-CI-features)(with minimal CI). And the result of new experiments that cover all cases looks good ([Azure Dashboard](https://dev.azure.com/ueqri-ci/pcl-test/_build?view=runs), from #20200420.12 to the latest, details please see appendix). Here's a short summary about my ideas:
![ci-main](https://user-images.githubusercontent.com/56567688/115573243-29faff80-a2f3-11eb-85c0-df9c77fdda76.png)

To control the CI workflow better and make the structure more perspicuous, I add a main pipeline to: 
1. **do formatting**, which will be done before both build & docs pipeline
2. **diff code changes** between two adjacent commits (to determine which pipeline, build or docs, should run next)

#### Pros
* remedy the unexpected behavior
* make the CI more explicit
* separate format stage, so that we can add stages and jobs as upstream of formatting in main pipeline
* easier to add diff rules (e.g. use wildcard or regex) in `git diff` script than using Azure path filter
#### Cons
* need to introduce another pipeline, which may be not a good choice for Azure Ops. (TBH, pipeline trigger is much smoother to maintain than the flow-control methods for stages or jobs)

<details>
 <summary><b>Appendix:</b> Details about experiment result</summary>

![Screenshot_2021-04-21_13-24-32](https://user-images.githubusercontent.com/56567688/115584835-adb9e980-a2fd-11eb-8f4f-3b1fb9f53437.png)

#### Test: only codes changed

Main Pipeline [buildID=160](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=160) -> Build [buildID=161](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=161) -> Docs [buildID=162](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=162)

#### Test: only docs changed

Main Pipeline [buildID=163](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=163) -> Docs [buildID=164](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=164)

#### Test: both codes and docs changed

Main Pipeline [buildID=165](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=165) -> Build [buildID=166](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=166) -> Docs [buildID=167](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=167)

#### Test: modified README.md

No CI was triggered.

</details>

ueqri commented:
The experiment branch is tested by some direct push currently. If it's possible to apply this new feature to our current CI, I'd change `git diff` command arguments to support for PR scenario. :blush: It's really a pleasure for me to work on the enhancement of CI!

kunaltyagi commented:
Sorry for taking a long time to reply. I went through your [CI file](https://github.com/ueqri/pcl/blob/7d359404e59fe56a54cdde9786ec0e659f1aa7ee/.ci/azure-pipelines/azure-pipelines.yaml), and the concept is nice.

However, I have a few questions:
* is there a way to keep dependencies only in 1 place (pipeline yaml or the script inside). Both places makes it brittle
* could you reduce the script to only the parts actually needed for info + actual functionality? That'd make it easier to decide if this complexity is helpful :)

ueqri commented:
Thanks for your kind reviews! I'm sorry for not updating the following comment in time. 
> The experiment branch is tested by some direct push currently. If it's possible to apply this new feature to our current CI, I'd change git diff command arguments to support for PR scenario. blush It's really a pleasure for me to work on the enhancement of CI!

The push trigger test is fine with [this CI file](https://github.com/ueqri/pcl/blob/443e01a629eaf25e040530b39314836fb566fa9a/.ci/azure-pipelines/azure-pipelines.yaml), and work well as follows:
<details>
 <summary><b>Appendix:</b> Details about first experiment result</summary>

![Screenshot_2021-04-21_13-24-32](https://user-images.githubusercontent.com/56567688/115584835-adb9e980-a2fd-11eb-8f4f-3b1fb9f53437.png)

#### Test: only codes changed

Main Pipeline [buildID=160](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=160) -> Build [buildID=161](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=161) -> Docs [buildID=162](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=162)

#### Test: only docs changed

Main Pipeline [buildID=163](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=163) -> Docs [buildID=164](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=164)

#### Test: both codes and docs changed

Main Pipeline [buildID=165](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=165) -> Build [buildID=166](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=166) -> Docs [buildID=167](https://dev.azure.com/ueqri-ci/pcl-test/_build/results?buildId=167)

#### Test: modified README.md

No CI was triggered.

</details>

The next PR test seems to make the git log and pipeline dashboard a bit messy, sorry for that. 

I used `test-CI-features` as a base branch and modified the pipeline to diff HEAD & `merge-base HEAD test-CI-features`. 
But I met with difficulties on the `git-diff` in Azure environment, and confused about that: the git command(e.g. `git diff HEAD...test-CI-features`) runs smoothly in my desktop with the same git version, but it doesn't work in Azure at all and throws fatal errors (e.g. `fatal: Not a valid object name test-CI-features`). So I use quite a lot scripts to debug, as you see in the last commit.

***

> is there a way to keep dependencies only in 1 place (pipeline yaml or the script inside). Both places makes it brittle

I'm not very sure about the "dependencies", could you give me a little hint or explanation? Should I just use `template` method to separate them into 1 place to keep the *-pipeline.yaml simplest?

> could you reduce the script to only the parts actually needed for info + actual functionality? That'd make it easier to decide if this complexity is helpful :)

Does the "script" refer to the script in code diff stage? If so, I'll definitely make it tidy after the test. Sorry for the current obscure situation about the code diff stage. And I'm grateful for any suggestions or hints to understand the git-diff error in CI.

kunaltyagi commented:
> I'm not very sure about the "dependencies"

I meant the files and folder to exclude. In the pipeline yaml file I linked to, the exclusion is mentioned in the yaml as well as in the diff script. It'd be nice to have to mention that only in 1 place

> Differences in azure and desktop git-diff

Azure does a raw copy and directly checks out the commit. Hence the difference in observable behavior.

> Does the "script" refer to the script in code diff stage? If so, I'll definitely make it tidy after the test

Yes. Please feel free to test, and report back. In the meantime, shall we mark this PR as draft?

PS: I'm sorry, I haven't worked with raw repositories as well, so I'll be of limited help there

ueqri commented:
Thanks for your help and advice! :grinning: 
I'd fix the dependencies of exclude rules, test git-diff for PR in Azure soon and report any milestones back.
> shall we mark this PR as draft?

It's really a good idea since the test hasn't been done. But this work has gone far from the topic "[CI] Fixed simple typo in doc/ to test docs-pipeline", should we move it to another draft PR, or rename this and mark it as draft?

kunaltyagi commented:
No need. You can use this PR to test the CI later once your git diff based pipeline changes are merged to PCL :)

ueqri commented:
> No need. You can use this PR to test the CI later once your git diff based pipeline changes are merged to PCL :)

Thanks for your help :blush: ! I'd go on the test and work it out soon.

ueqri commented:
Thank you for your guidance and suggestions in this PR!  :grinning: 

I've fixed the git diff based pipeline, and the test on PR from another user's branch in a forked repo to my branch `test-CI-features` is finished successfully. 

The result is here:
- [Azure Dashboard](https://dev.azure.com/ueqri-ci/pcl-test/_build?view=runs)
- [PR list in my forked repo](https://github.com/ueqri/pcl/pulls)

![Screenshot_2021-04-29_02-05-25](https://user-images.githubusercontent.com/56567688/116451711-6c8e8000-a88f-11eb-89b2-22f5f1692ff5.png)

And I applied the changes for pipeline in a new branch `CI-git-diff-based`, it's all ready now. :blush:

PS: For dependencies of exclude rules, it's a bit hard to separate the path-filter(a part of Azure YAML schema), and the diff-exclude-rules to one place. So as a second-best solution, I pick up the diff-exclude-rules to the head of the script lines, and add some comments to attract attention in YAML files.

```yaml
# Exclude Rules:
# 1. path-filter in trigger and pr, 
#    this is NOT exclude rules for documentation,
#    set this rule only to avoid any CI run.
# 2. DOCS_LISTS in diff script, 
#    the lists contains all documentation-related files,
#    fill the list to help git-diff recognize docs changes.
```
https://github.com/ueqri/pcl/blob/CI-git-diff-based/.ci/azure-pipelines/azure-pipelines.yaml#L48-L52

```yaml
...
- script: | 
     # Add all docs path to this array
     DOCS_LISTS=("doc" "*.md")
     ...
```
https://github.com/ueqri/pcl/blob/CI-git-diff-based/.ci/azure-pipelines/azure-pipelines.yaml#L1-L7
